src/org/omegat/core/machinetranslators/MTHUBTranslate.java

```diff
@@ -0,0 +1,412 @@
+            @Override
+            protected void onConfirm() {
+                String key = panel.valueField1.getText().trim();
+                boolean temporary = panel.temporaryCheckBox.isSelected();
+                setCredential(PROPERTY_API_KEY, key, temporary);
+            }
+        };
+        dialog.panel.valueLabel1.setText(OStrings.getString(
+                                              "MT_ENGINE_MTHUB_API_KEY_LABEL"));
+        dialog.panel.valueField1.setText(getCredential(PROPERTY_API_KEY));
+        dialog.panel.valueLabel2.setVisible(false);
+        dialog.panel.valueField2.setVisible(false);
+        dialog.panel.temporaryCheckBox.setSelected(
+                isCredentialStoredTemporarily(PROPERTY_API_KEY));
+        dialog.show();
+    }
+
+    @Override
+    protected String translate(Language sLang, Language tLang, String text)
+            throws Exception {
+
+        String mthubKey = getCredential(PROPERTY_API_KEY);
+        if (mthubKey == null || mthubKey.isEmpty()) {
+            return OStrings.getString("MT_ENGINE_MTHUB_KEY_NOTFOUND");
+        }
+
+        String prev = getFromCache(sLang, tLang, text);
+        if (prev != null) {
+            return prev;
+        }
+
+        if (this.availableLanguageCodes == null) {
+            this.getAvailableLanguageCodes();
+            if (this.availableLanguageCodes == null) {
+                /*error to obtain the available language,
+                it is a error with the user key*/
+                return OStrings.getString("MT_ENGINE_MTHUB_INVALID_KEY");
+            }
+        }
+
+        String encoding = StandardCharsets.UTF_8.name();
+        try {
+            URLConnection connection = new URL(GT_URL).openConnection();
+            connection.setDoOutput(true);
+            connection.setRequestProperty("Accept-Charset", encoding);
+            connection.setRequestProperty("Content-Type",
+                                        "application/json;charset=" + encoding);
+
+            try (OutputStream output = connection.getOutputStream()) {
+
+                OM.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY,
+                        true);
+                List<String> segments = Collections.singletonList(text.substring(0,
+                                 Math.min(text.length(), LIMIT_CHARACTER - 1)));
+
+                MTHUBTranslateRequest mttr = new MTHUBTranslateRequest(
+                        mthubKey, normaliseCode(sLang), normaliseCode(tLang),
+                        segments);
+
+                String json = OM.writeValueAsString(mttr);
+                output.write(json.getBytes(encoding));
+
+                MTHUBTranslateResponse response = OM.readValue(
+                                                  connection.getInputStream(),
+                                                  MTHUBTranslateResponse.class);
+
+                String translation = "";
```

Comment:
The way you are handling `translation` here, if the response is not success then you will cache the empty string as the translation. Are you sure that's what you want to do?

Reply:
I changed it, and it only set in cache when the response is success